core/lib/zk_os_merkle_tree/src/hasher/nodes.rs

```diff
@@ -0,0 +1,183 @@
+                    )
+                };
+            }
+            hashes.truncate(new_len);
+            on_level(&hashes);
+        }
+
+        hashes[0]
+    }
+
+    fn internal_hashes<P: TreeParams>(&self, hasher: &P::Hasher, depth: u8) -> InternalNodeHashes {
+        // capacity = 2 + 4 + ... + 2 ** (P::INTERNAL_NODE_DEPTH - 1) = 2 * (2 ** (P::INTERNAL_NODE_DEPTH - 1) - 1) = 2 ** P::INTERNAL_NODE_DEPTH - 2
+        let capacity = (1 << P::INTERNAL_NODE_DEPTH) - 2;
+        let mut hashes = InternalNodeHashes(Vec::with_capacity(capacity));
+        let mut full_level_len = 1 << (P::INTERNAL_NODE_DEPTH - 1);
+        self.hash_inner::<P>(hasher, depth, false, |level_hashes| {
+            hashes.0.extend_from_slice(level_hashes);
+            // Pad if necessary so that there are uniform offsets for each level. The padding should never be read.
+            // It doesn't waste that much space given that it may be required only for one internal node per level.
+            hashes
+                .0
+                .extend(iter::repeat(H256::zero()).take(full_level_len - level_hashes.len()));
+            full_level_len /= 2;
+        });
+        assert_eq!(hashes.0.len(), capacity);
+        hashes
+    }
+}
+
+impl Root {
+    pub(crate) fn hash<P: TreeParams>(&self, hasher: &P::Hasher) -> H256 {
+        self.root_node.hash::<P>(
+            hasher,
+            max_nibbles_for_internal_node::<P>() * P::INTERNAL_NODE_DEPTH,
+        )
+    }
+}
+
+/// Internal hashes for a level of `InternalNode`s.
+#[derive(Debug)]
+pub(crate) struct InternalHashes<'a> {
+    nodes: &'a HashMap<u64, InternalNode>,
+    // TODO: `Vec<(u64, H256)>` for a level may be more efficient
+    /// Ordered by ascending depth `1..internal_node_depth` where `depth == 1` is just above child refs.
+    internal_hashes: HashMap<u64, InternalNodeHashes>,
+    internal_node_depth: u8,
+    level_offsets: Vec<usize>,
+}
+
+impl<'a> InternalHashes<'a> {
+    pub(crate) fn new<P: TreeParams>(
+        nodes: &'a HashMap<u64, InternalNode>,
+        hasher: &P::Hasher,
+        depth: u8,
+    ) -> Self {
+        use rayon::prelude::*;
+
+        let mut offset = 0;
+        let mut level_offsets = Vec::with_capacity(usize::from(P::INTERNAL_NODE_DEPTH) - 1);
+        for depth_in_node in 1..P::INTERNAL_NODE_DEPTH {
+            level_offsets.push(offset);
+            offset += 1_usize << (P::INTERNAL_NODE_DEPTH - depth_in_node);
+        }
+
+        let internal_hashes = nodes
+            .par_iter()
+            .map(|(idx, node)| (*idx, node.internal_hashes::<P>(hasher, depth)))
+            .collect();
+
+        Self {
+            nodes,
+            internal_hashes,
+            internal_node_depth: P::INTERNAL_NODE_DEPTH,
+            level_offsets,
+        }
+    }
+
+    pub(crate) fn get(&self, depth_in_node: u8, index_on_level: u64) -> H256 {
+        let bit_shift = self.internal_node_depth - depth_in_node;
```

Comment:
Does it make sense to validate `depth_in_node` and `index_on_level`? i.e add asserts
`assert(depth_in_node < self.internal_node_depth)`
`assert(index_on_level < self.nodes.len() * (1 << bit_shift))`

Reply:
It's possible, but I'm reasonably sure that both assertions hold by construction due to how `get()` is used in `collect_hashes()`. E.g., `depth_in_node` is defined as `depth % P::INTERNAL_NODE_DEPTH`, i.e., it's always valid.